### ISSUE TITLE
"Copy Link" and "Copy Text" commands for a post do not work with server v7.10.3

### ISSUE THREAD

**Issue by gucchy55** [mattermost/mattermost#23978]
#### Summary
"Copy Link" and "Copy Text" command for a post does not work.

#### Steps to reproduce
Server: v7.10.3 on Ubuntu
Client: v5.4.0 (on both Windows and Mac)
"Copy Link" and "Copy Text" commands do not put the post content into the clipboard.
It was no problem with server v7.9.1 and the same client version.

#### Expected behavior
The commands get to work again.

**Comment by gucchy55** [comment]
Thank you. In my environment, the copy commands do not put anything into the clipboard, which looks a different case from that of the mentioned ticket.
Anyway, I will try with v8.0 when it is released.

**Comment by amyblais** [comment]
@gucchy55 Thank you, can you help share the commands you're testing with to help me reproduce this?

**Comment by gucchy55** [comment]
With a desktop app (Windows and Mac), even when I select the menu "Copy Link/Text" for a post, it does not put anything into the clipboard. Is it an issue of the desktop app?

**Comment by amyblais** [comment]
I wasn't able to reproduce it on v7.10.4 on Mac either on webapp or desktop app. If you're open to it, please help create an account at https://community.mattermost.com to see if the issue reproduces for you there. This is our active community server and on the latest version.

**Comment by gucchy55** [comment]
I have roll-backed the server to v7.9.5, and it has no problem now. I will re-try the future version again. Thank you.

**Comment by f-velka** [comment]
I have the same problem, and was able to reproduce it on Windows.

1. Launch mattermost v7.10.4 team edition using https://github.com/mattermost/docker on localhost:8065
2. Add the entry  `127.0.0.1 mattermost.example.com` to the hosts file
3. Access mattermost via `http://mattermost.example.com:8065` on browser
4. Try "Copy Link" or "Copy Text"

"Copy Link" and "Copy Text" work when I access via `http://localhost:8065`.
I think the [copyToClipboard](https://github.com/mattermost/mattermost/blob/40ff77c9c6c2148e33a7fc5f49441b00bffd5c4c/webapp/channels/src/utils/utils.tsx#L1512) function isn't working properly when the context isn't secure and the Clipboard API is disabled.

**Comment by gucchy55** [comment]
That is the case. I have a http environment (without ssl) internally.

**Comment by amyblais** [comment]
Thanks, opened https://mattermost.atlassian.net/browse/MM-53680.

**Comment by devinbinnie** [comment]
This is actually working as designed as per the design of the Clipboard API (see here: https://developer.mozilla.org/en-US/docs/Web/API/Clipboard). You will need to use HTTPS in order to work with the clipboard features we've added in Mattermost.

**Comment by f-velka** [comment]
@devinbinnie 

As you mentioned, https is needed to use the Clipboard API, but in the previous versions such as v7.9.5, the "Copy Link" and "Copy Text" functions did work even with http, and it was thanks to the workaround below in [copyToClipboard](https://github.com/mattermost/mattermost/blob/40ff77c9c6c2148e33a7fc5f49441b00bffd5c4c/webapp/channels/src/utils/utils.tsx#L1512), maybe.

```TypeScript
    // creates a tiny temporary text area to copy text out of
    // see https://stackoverflow.com/a/30810322/591374 for details
    const textArea = document.createElement('textarea');
    textArea.style.position = 'fixed';
    textArea.style.top = '0';
    textArea.style.left = '0';
    textArea.style.width = '2em';
    textArea.style.height = '2em';
    textArea.style.padding = '0';
    textArea.style.border = 'none';
    textArea.style.outline = 'none';
    textArea.style.boxShadow = 'none';
    textArea.style.background = 'transparent';
    textArea.value = data;
    document.body.appendChild(textArea);
    textArea.select();
    document.execCommand('copy');
    document.body.removeChild(textArea);
```
But even that stopped working in v7.10.4.

I suspect that making or selecting the temporary text area before `document.execCommand('copy')` isn't working well, because when I select some text by dragging and then click "Copy Text", the selected text gets copied to the clipboard.

**Comment by syockit** [comment]
While I can understand the rationale for not enabling clipboard over unsecured internet, it makes less sense for intranet (where usually https is not enabled) (also including VPN). Can we at least have it enabled for the desktop version?

**Comment by CJCombrink** [comment]
@devinbinnie 
This is unfortunate for a server that is in an intranet.
Is it possible to at least put something in the clipboard or show an error indicating the issue?
Something like: "Copy not supported for http servers"
Not copying anything and no indication of the issue also feels like a bug.

**Comment by syockit** [comment]
I wrote an Edge extension that injects a script to create an empty object to replace `navigator.clipboard`, and appended a function to its `writeText` method that does the above workaround as posted by @f-velka, and it still didn't work. After some fiddling, I found that perhaps the textArea didn't get enough time to actually spawn for its `select()` method to properly work, so I enclosed everything after `appendChild` inside a `setTimeout` (with 100ms delay). I've also set `style.zIndex = 1` and `style.display = "block"`, thought I'm not sure if they were necessary. Anyways, I guess enclosing that's what's needed to fix the code for the time being (until execCommand is removed).
